### DISCUSSION TITLE
support both http1, http2, and https

### DISCUSSION THREAD

**Discussion by rootBiao** [actix/actix-web#3084] [Q&A] [answered]
Is there any way to enable Actix-web to support both http1, http2, and https?

**Comment by robjtede** [comment, marked as answer]
When TLS/HTTPS is set up, both HTTP/1.x and HTTP/2 are supported.

https://actix.rs/docs/http2

**Reply by DeveshSharma2004** [reply]
Actix web support TLS 1.3?

**Comment by rootBiao** [comment]
Is there a way to have http1, http2, and https listen to different ports separately？

**Reply by robjtede** [reply]
Yes, you can call `.bind()` and `.bind_rustls()` on the same `HttpServer`.

**Reply by rootBiao** [reply]
What methods are there to use http2 without using https, such as listening to port 8080 for http1, listening to port 80 for http2, and listening to port 443 for https

**Reply by robjtede** [reply]
HTTP/2 without TLS is implemented and will be released in the upcoming v4.4 release. At that point you'll be able to do this to support your use case:

```rust
  .bind(8080)? // <-- only HTTP/1.x
  .bind_auto_h2c(80)? // <-- HTTP/1.x & HTTP/2
  .bind_rustls(443, ...)? // <-- HTTP/1.x & HTTP/2 (over TLS)
```

**Reply by rootBiao** [reply]
In lower versions of Actix-web, http1 and http2 can listen to the same port, such as 80. Will this feature be supported in future versions？

**Reply by robjtede** [reply]
It's not currently possible without using the lower level `actix-http` crate.

Currently:

```rust
  .bind(8080)? // <-- only HTTP/1.x
  .bind_rustls(443, ...)? // <-- HTTP/1.x & HTTP/2 (over TLS)
```
